File: backend/utils/calculations.py
def calculate_scores(solution, scores, criteria):
    # Group criteria by category
    categories = {}
    
    # Initialize categories
    for c in criteria:
        cat_name = c['category']
        if cat_name not in categories:
            categories[cat_name] = {
                'name': cat_name,
                'totalScore': 0,
                'maxPossibleScore': 0,
                'weightedTotal': 0,
                'weightSum': 0,
                'criteria': []
            }
        categories[cat_name]['criteria'].append(c)

    # Process scores
    for score_item in scores:
        criterion = next((c for c in criteria if c['id'] == score_item['criterionId']), None)
        if criterion:
            cat = categories[criterion['category']]
            try:
                numeric_score = float(score_item['score'])
            except:
                numeric_score = 0
            
            # Raw score sum
            cat['totalScore'] += numeric_score
            # Max potential (assuming 10 is max score)
            cat['maxPossibleScore'] += 10
            
            # Weighted calculation
            weight = criterion.get('weight', 1)
            cat['weightedTotal'] += numeric_score * weight
            cat['weightSum'] += weight

    # Calculate averages per category
    category_results = []
    for cat in categories.values():
        avg_score = (cat['weightedTotal'] / cat['weightSum']) if cat['weightSum'] > 0 else 0
        category_results.append({
            'category': cat['name'],
            'score': round(float(avg_score), 2),
            'rawTotal': cat['totalScore'],
            'itemCount': len(cat['criteria'])
        })

    # Calculate Overall Score
    overall_total = sum(cat['score'] for cat in category_results)
    overall_score = (overall_total / len(category_results)) if category_results else 0

    # MITRE ATT&CK Coverage
    mitre_criteria = [c for c in criteria if c['category'] == 'Detection & Prevention']
    mitre_scores = [s for s in scores if any(c['id'] == s['criterionId'] for c in mitre_criteria)]
    
    mitre_score_total = sum(float(s['score']) for s in mitre_scores if str(s['score']).replace('.','',1).isdigit())
    mitre_max_total = len(mitre_criteria) * 10
    
    coverage_percentage = (mitre_score_total / mitre_max_total * 100) if mitre_max_total > 0 else 0

    return {
        'solutionId': solution['id'],
        'solutionName': solution['name'],
        'overallScore': round(overall_score, 2),
        'categoryScores': category_results,
        'metrics': {
            'mitreCoverage': f"{round(coverage_percentage, 1)}%",
            'falsePositiveRate': 'Low', # Placeholder
            'costEfficiency': 'Medium'  # Placeholder
        }
    }

def run_comparison(solutions, all_scores, criteria):
    """
    Compare all solutions based on their scores.
    
    Args:
        solutions: List of solution objects with 'id', 'name', etc.
        all_scores: List of score records, each with 'solutionId', 'timestamp', and 'items'
        criteria: List of criteria objects
    
    Returns:
        Dictionary with 'rankings' and 'categoryAverages'
    """
    results = []
    
    for solution in solutions:
        # Find all score records for this solution
        solution_score_records = [s for s in all_scores if s.get('solutionId') == solution['id']]
        
        # Extract the items from the most recent score record
        # Flatten all score items from all records for this solution
        solution_score_items = []
        for record in solution_score_records:
            items = record.get('items', [])
            solution_score_items.extend(items)
        
        if solution_score_items:
            result = calculate_scores(solution, solution_score_items, criteria)
            
            # Add additional metrics from solution data
            result['metrics']['deploymentType'] = solution.get('deploymentModel', 'N/A')
            result['metrics']['vendor'] = solution.get('vendor', 'N/A')
            result['metrics']['annualCost'] = solution.get('cost', 0)
            
            # Convert category scores list to a dictionary for easier frontend access
            category_score_dict = {}
            for cat in result.get('categoryScores', []):
                category_score_dict[cat['category']] = cat['score']
            result['categoryScores'] = category_score_dict
            
            results.append(result)

    # Sort by overall score descending
    results.sort(key=lambda x: x.get('overallScore', 0), reverse=True)

    # Calculate category averages across all solutions
    category_averages = {}
    if results:
        # Get all unique categories
        all_categories = set()
        for res in results:
            all_categories.update(res.get('categoryScores', {}).keys())
        
        # Calculate average for each category
        for category in all_categories:
            scores = [res['categoryScores'].get(category, 0) for res in results if category in res.get('categoryScores', {})]
            if scores:
                category_averages[category] = round(sum(scores) / len(scores), 2)

    return {
        'rankings': results,
        'categoryAverages': category_averages,
        'timestamp': None  # Add timestamp if needed
    }

File: backend/utils/calculations.js
// Utility functions for calculating scores and generating reports

/**
 * Calculate detailed scores for a single solution
 */
const calculateScores = (solution, scores, criteria) => {
    // Group criteria by category
    const categories = {};
    const totalWeights = {}; // Sum of weights per category (for weighted average)

    // Initialize categories map
    criteria.forEach(c => {
        if (!categories[c.category]) {
            categories[c.category] = {
                name: c.category,
                totalScore: 0,
                maxPossibleScore: 0,
                weightedTotal: 0,
                weightSum: 0,
                criteria: []
            };
        }
        categories[c.category].criteria.push(c);
    });

    // Process scores
    scores.forEach(scoreItem => {
        const criterion = criteria.find(c => c.id === scoreItem.criterionId);
        if (criterion) {
            const cat = categories[criterion.category];
            const numericScore = parseFloat(scoreItem.score) || 0;

            // Raw score sum
            cat.totalScore += numericScore;
            // Max potential (assuming 10 is max score)
            cat.maxPossibleScore += 10;

            // Weighted calculation
            const weight = criterion.weight || 1;
            cat.weightedTotal += numericScore * weight;
            cat.weightSum += weight;
        }
    });

    // Calculate averages per category
    const categoryResults = Object.values(categories).map(cat => {
        const avgScore = cat.weightSum > 0 ? (cat.weightedTotal / cat.weightSum) : 0;
        return {
            category: cat.name,
            score: parseFloat(avgScore.toFixed(2)),
            rawTotal: cat.totalScore,
            itemCount: cat.criteria.length // This might need to be 'answered items' count in a real scenario
        };
    });

    // Calculate Overall Score (Simple average of category scores for now, or weighted category avg)
    // Let's do a weighted average of categories if we had category weights, but for now simple average of category scores
    const overallTotal = categoryResults.reduce((sum, cat) => sum + cat.score, 0);
    const overallScore = categoryResults.length > 0 ? (overallTotal / categoryResults.length) : 0;

    // MITRE ATT&CK Coverage
    // Filter related criteria
    const mitreCriteria = criteria.filter(c => c.category === 'Detection & Prevention');
    const mitreScores = scores.filter(s => {
        const c = criteria.find(cri => cri.id === s.criterionId);
        return c && c.category === 'Detection & Prevention';
    });

    let mitreScoreTotal = 0;
    let mitreMaxTotal = mitreCriteria.length * 10;

    mitreScores.forEach(s => mitreScoreTotal += (parseFloat(s.score) || 0));
    const coveragePercentage = mitreMaxTotal > 0 ? ((mitreScoreTotal / mitreMaxTotal) * 100) : 0;

    return {
        solutionId: solution.id,
        solutionName: solution.name,
        overallScore: parseFloat(overallScore.toFixed(2)),
        categoryScores: categoryResults,
        metrics: {
            mitreCoverage: parseFloat(coveragePercentage.toFixed(1)) + '%',
            // Add other mock metrics or calculated ones here
            falsePositiveRate: 'Low', // Placeholder as it depends on specific criteria input
            costEfficiency: 'Medium'
        }
    };
};

/**
 * Compare multiple solutions
 */
const runComparison = (solutions, allScores, criteria) => {
    const results = solutions.map(solution => {
        const solutionEntry = allScores.find(s => s.solutionId === solution.id);
        const scores = solutionEntry ? solutionEntry.items : [];
        return calculateScores(solution, scores, criteria);
    });

    // Sort by overall score descending
    results.sort((a, b) => b.overallScore - a.overallScore);

    // Calculate Category Averages
    const categoryTotals = {};
    const categoryCounts = {};

    results.forEach(sol => {
        sol.categoryScores.forEach(c => {
            if (!categoryTotals[c.category]) {
                categoryTotals[c.category] = 0;
                categoryCounts[c.category] = 0;
            }
            categoryTotals[c.category] += c.score;
            categoryCounts[c.category]++;
        });
    });

    const categoryAverages = {};
    Object.keys(categoryTotals).forEach(cat => {
        categoryAverages[cat] = parseFloat((categoryTotals[cat] / categoryCounts[cat]).toFixed(2));
    });

    return {
        rankings: results,
        categoryAverages,
        timestamp: new Date().toISOString()
    };
};

module.exports = {
    calculateScores,
    runComparison
};
